python/tritonserver/_api/_allocators.py

```diff
@@ -0,0 +1,323 @@
+        memory type id (typically the same as device id)
+    size : int
+        Size of the allocated memory in bytes.
+    owner : Any
+        Object that owns or manages the memory buffer.  Allocated
+        memory must not be freed while a reference to the owner is
+        held.
+
+    Examples
+    --------
+    >>> buffer = MemoryBuffer.from_dlpack(numpy.array([100],dtype=numpy.uint8))
+
+    """
+
+    data_ptr: int
+    memory_type: MemoryType
+    memory_type_id: int
+    size: int
+    owner: Any
+
+    @staticmethod
+    def from_dlpack(owner: Any) -> MemoryBuffer:
+        if not hasattr(owner, "__dlpack__"):
+            raise InvalidArgumentError("Object does not support DLpack protocol")
+
+        dlpack_object = DLPackObject(owner)
+
+        if not dlpack_object.contiguous:
+            raise InvalidArgumentError("Only contiguous memory is supported")
+
+        return MemoryBuffer(
+            int(dlpack_object.data_ptr),
+            dlpack_object.memory_type,
+            dlpack_object.memory_type_id,
+            dlpack_object.byte_size,
+            owner,
+        )
+
+    @staticmethod
+    def _from_dlpack_object(owner: Any, dlpack_object: DLPackObject) -> MemoryBuffer:
+        if not dlpack_object.contiguous:
+            raise InvalidArgumentError("Only contiguous memory is supported")
+
+        return MemoryBuffer(
+            int(dlpack_object.data_ptr),
+            dlpack_object.memory_type,
+            dlpack_object.memory_type_id,
+            dlpack_object.byte_size,
+            owner,
+        )
+
+    def _create_TRITONSERVER_BufferAttributes(self) -> TRITONSERVER_BufferAttributes:
+        buffer_attributes = TRITONSERVER_BufferAttributes()
+        buffer_attributes.memory_type = self.memory_type
+        buffer_attributes.memory_type_id = self.memory_type_id
+        buffer_attributes.byte_size = self.size
+        #        buffer_attributes.cuda_ipc_handle = None
+        return buffer_attributes
+
+
+class MemoryAllocator(ABC):
+    """Abstract interface to allow for custom memory allocation strategies
+
+    Classes implementing the MemoryAllocator interface have to provide
+    an allocate method returning MemoryBuffer objects.  A memory
+    allocator implementation does not need to match the requested
+    memory type or memory type id.
+
+
+
+    Examples
+    --------
+
+    class TorchAllocator(tritonserver.MemoryAllocator):
+        def allocate(self,
+                     size,
+                     memory_type,
+                     memory_type_id,
+                     tensor_name):
+
+            device = "cpu"
+
+            if memory_type == tritonserver.MemoryType.GPU:
+                device = "cuda"
+
+            tensor = torch.zeros(size,dtype=torch.uint8,device=device)
+            return tritonserver.MemoryBuffer.from_dlpack(tensor)
+
+    """
+
+    @abstractmethod
+    def allocate(
+        self, size: int, memory_type: MemoryType, memory_type_id: int, tensor_name: str
```

Comment:
Attaching comments on removing `tensor_name` here since code has changed: https://github.com/triton-inference-server/core/pull/299#discussion_r1446484855